### PR TITLE
fix: add shim for `require('node:buffer')`

### LINKED ISSUE #2047: Add shim for Buffer
Next.js [13.3.1-canary.8](https://www.npmjs.com/package/next/v/13.3.1-canary.8) introduces a [change to the fetch handling](https://github.com/vercel/next.js/pull/48365/files#diff-4c54e369ddb9a2db1eed95fe1d678f94c8e82c540204475d42c78e49bf4f223aR245) that uses the global Buffer object. This causes a runtime error because Buffer is not available in Deno.

@Skn0tt has commented on the PR, but until this is fixed (or assuming it won't be) we should implement a workaround by adding Buffer to [our shims](https://github.com/netlify/next-runtime/blob/main/packages/runtime/src/templates/edge/shims.js).

### PR DESCRIPTION
**Pull request by orinokai**
### Summary

A new turbopack feature allows free variables to be aliased to modules, which causes a Next.js reference to `Buffer` to be bundled as `require("node:buffer")`, which is not supported by Deno.

This PR adds a shim for require that specifically loads `Buffer` for this case only.

### Test plan

Incoming...

### Relevant links (GitHub issues, Notion docs, etc.) or a picture of cute animal

Fixes #2047

### CODE REVIEW

**Comment by ascorbic** [comment]
I think adding `require` potentially opens up a world of pain. I don't get why it's generating a cjs require anyway. Has Deno added support for `"node:` imports to Deploy yet? It might be better to patch the bundle instead.

**Comment by pieh** [comment]
Just adding https://github.com/vercel/next.js/pull/47191/files#diff-be23f4ad09849e271e6c9f4f373142bc18167eb9f4fe6a86f8705b80e07995d3R202-R211 for some reference - I'm not sure of where `server/sandbox` is used, but `require` function is added there to context

Also above explains how we end up with that require (externals -> commonjs in webpack config) and also gives potential hint that if we don't want to produce code with `require` calls we would need to overwrite default `externals` config and supply alternative external for `buffer` (and other Node native things from above PR). If there is non-hacky way to do that I think that would be preferable to patching produced bundle after the fact? (I think you could do that in `next.config.js` in bundle configuration overwrites, but mutating config from build plugin/runtime is generally a mess :) )

**Comment by ascorbic** [comment]
What a car crash

**Comment by orinokai** [comment]
does anyone have any insight into why Vercel pick specific properties out of the native modules? https://github.com/vercel/next.js/pull/47191/files#diff-be23f4ad09849e271e6c9f4f373142bc18167eb9f4fe6a86f8705b80e07995d3R149-R156

**Comment by nickytonline** [comment]
> does anyone have any insight into why Vercel pick specific properties out of the native modules? [vercel/next.js#47191 (files)](https://github.com/vercel/next.js/pull/47191/files#diff-be23f4ad09849e271e6c9f4f373142bc18167eb9f4fe6a86f8705b80e07995d3R149-R156)

No idea why they omitted some.

**Comment by ascorbic** [comment]
This is why! https://github.com/cloudflare/workerd/blob/main/src/node/buffer.ts

**Review by LekoArts — APPROVED** [review]

**Comment by ascorbic** [comment]
Are you planning to add shims for the other stuff too?

File: packages/runtime/src/templates/edge/shims.js [hunk 1]
```diff
@@ -48,6 +49,16 @@ const fetch /* type {typeof globalThis.fetch} */ = async (url, init) => {
   }
 }
 
+// Turbopack aliases "Buffer" to a module import, so we need to provide a shim for that
+if (typeof require === 'undefined') {
+  globalThis.require = (name) => {
+    if (name === 'buffer' || name === 'node:buffer') {
+      return { Buffer: BufferCompat }
+    }
+    throw new ReferenceError('require is not defined')
```
[the remark this answers is not in the document]
Comment (nickytonline) [reply packages/runtime/src/templates/edge/shims.js:58]: Even better 😎

[the remark this answers is not in the document]
Comment (orinokai) [reply packages/runtime/src/templates/edge/shims.js:58]: > Are you planning to add shims for the other stuff too?

Do you mean the other stuff in the [`NativeModuleMap`](https://github.com/vercel/next.js/pull/47191/files#diff-be23f4ad09849e271e6c9f4f373142bc18167eb9f4fe6a86f8705b80e07995d3R147-R185), @ascorbic? I wasn't certain whether that was specific to Vercel's infrastructure?

[the remark this answers is not in the document]
Comment (ascorbic) [reply packages/runtime/src/templates/edge/shims.js:58]: They're all the Node modules that are available in workerd, so which Vercel edge runtime now assumes are available. Buffer is the only on they're using in core Next.js right now (I think), but they're basically telling users that they can use them.

File: packages/runtime/src/templates/edge/shims.js [hunk 2]
```diff
@@ -48,6 +52,26 @@ const fetch /* type {typeof globalThis.fetch} */ = async (url, init) => {
   }
 }
 
+// Shim native modules that Vercel makes available
+if (typeof require === 'undefined') {
```
Comment (nickytonline) [inline packages/runtime/src/templates/edge/shims.js:56]: So many shims. 😅

**Review by nickytonline — APPROVED** [review]
Looks great @orinokai!  :shipit:

**Review by LekoArts — APPROVED** [review]

**Pull request merged**